telemetry/derive/src/lib.rs

```diff
@@ -145,54 +142,91 @@ impl ToTokens for MetricSpec {
 ///     Ok(())
 /// }
 /// ```
-#[proc_macro_error]
+#[manyhow]
 #[proc_macro_attribute]
 pub fn metrics(attr: TokenStream, item: TokenStream) -> TokenStream {
-    let ItemFn {
+    let mut emitter = Emitter::new();
+
+    let Some(func): Option<syn2::ItemFn> = emitter.handle(syn2::parse2(item)) else {
+        return emitter.finish_token_stream();
+    };
+
+    // This is a good sanity check. Possibly redundant.
+    if func.sig.ident != "execute" {
+        emit!(
+            emitter,
+            func.sig.ident,
+            "Function should be an `impl execute`"
+        );
+    }
+
+    if func.sig.inputs.is_empty() {
+        emit!(
+            emitter,
+            func.sig,
+            "Function must have at least one argument of type `WorldStateView`."
+        );
+    }
+
+    let Some(metric_specs): Option<MetricSpecs> = emitter.handle(syn2::parse2(attr)) else {
+        return emitter.finish_token_stream();
+    };
+
+    let result = impl_metrics(&mut emitter, &metric_specs, &func);
+
+    emitter.finish_token_stream_with(result)
+}
+
+fn impl_metrics(emitter: &mut Emitter, attr: &MetricSpecs, func: &syn2::ItemFn) -> TokenStream {
+    let syn2::ItemFn {
         attrs,
         vis,
         sig,
         block,
-    }: ItemFn = parse_macro_input!(item as ItemFn);
+    } = func;
 
-    // This is a good sanity check. Possibly redundant.
-    if sig.ident != "execute" {
-        abort!(sig.ident, "Function should be an `impl execute`");
-    }
     match sig.output.clone() {
-        syn::ReturnType::Default => abort!(
+        syn2::ReturnType::Default => emit!(
+            emitter,
             sig.output,
             "`Fn` must return `Result`. Returns nothing instead. "
         ),
         #[allow(clippy::string_to_string)]
-        syn::ReturnType::Type(_, typ) => match *typ {
+        syn2::ReturnType::Type(_, typ) => match *typ {
             Type::Path(pth) => {
                 let Path { segments, .. } = pth.path;
                 let type_name = &segments.last().expect("non-empty path").ident;
                 if *type_name != "Result" {
-                    abort!(
+                    emit!(
+                        emitter,
                         type_name,
-                        format!("Should return `Result`. Found {type_name}")
+                        "Should return `Result`. Found `{:?}`",
```

Comment:
Shouldn't it be `{}`? At least it was the format used before..

Reply:
Yeah, changed it